File: src/test/java/macro303/keschet/Direction_Test.java
package macro303.keschet;

import macro303.board_game.Board;
import macro303.board_game.Colour;
import macro303.board_game.Coordinates;
import macro303.board_game.Square;
import macro303.keschet.pieces.General;
import macro303.keschet.pieces.Piece;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Created by Macro303 on 2018-02-08.
 */
public class Direction_Test {
	private static final Logger LOGGER = LogManager.getLogger();
	private static Board board;
	private static Square start;
	private static Square end;

	@BeforeAll
	static void beforeAll() {
		board = new Board(10);
		start = board.getSquare(new Coordinates(2, 2));
	}

	@BeforeEach
	void beforeEach(){
		start.setItem(new General(Colour.BLUE));
		if (end != null)
			end.setItem(null);
	}

	@Test
	void test_northDirection() {
		end = board.getSquare(new Coordinates(2, 1));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("North ==> " + direction);
		assertEquals(Direction.NORTH, direction);
	}

	@Test
	void test_northEastDirection() {
		end = board.getSquare(new Coordinates(3, 1));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("North-East ==> " + direction);
		assertEquals(Direction.NORTH_EAST, direction);
	}

	@Test
	void test_eastDirection() {
		end = board.getSquare(new Coordinates(3, 2));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("East ==> " + direction);
		assertEquals(Direction.EAST, direction);
	}

	@Test
	void test_southEastDirection() {
		end = board.getSquare(new Coordinates(3, 3));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("South-East ==> " + direction);
		assertEquals(Direction.SOUTH_EAST, direction);
	}

	@Test
	void test_southDirection() {
		end = board.getSquare(new Coordinates(2, 3));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("South ==> " + direction);
		assertEquals(Direction.SOUTH, direction);
	}

	@Test
	void test_southWestDirection() {
		end = board.getSquare(new Coordinates(1, 3));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("South-West ==> " + direction);
		assertEquals(Direction.SOUTH_WEST, direction);
	}

	@Test
	void test_westDirection() {
		end = board.getSquare(new Coordinates(1, 2));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("West ==> " + direction);
		assertEquals(Direction.WEST, direction);
	}

	@Test
	void test_northWestDirection() {
		end = board.getSquare(new Coordinates(1, 1));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("North-West ==> " + direction);
		assertEquals(Direction.NORTH_WEST, direction);
	}

	@Test
	void test_invalidDirection() {
		end = board.getSquare(new Coordinates(4, 3));
		assert end != null;
		Direction direction = Util.calculateDirection(start, end);
		LOGGER.info("Invalid ==> " + direction);
		assertEquals(Direction.INVALID, direction);
	}

	@Test
	void test_noDirection() {
		Direction direction = Util.calculateDirection(start, start);
		LOGGER.info("None ==> " + direction);
		assertEquals(Direction.INVALID, direction);
	}
}

File: Base/src/test/kotlin/macro303/base/Direction_Test.kt
package macro303.base

import org.junit.Test

/**
 * Created by Macro303 on 2018-02-07.
 */
class Direction_Test {

	@Test
	fun test_northDirection() {
		val north = IBoard.calculateDirection(Pair(2, 2), Pair(2, 1))
		assert(north == Direction.NORTH)
	}

	@Test
	fun test_northEastDirection() {
		val northEast = IBoard.calculateDirection(Pair(2, 2), Pair(3, 1))
		assert(northEast == Direction.NORTH_EAST)
	}

	@Test
	fun test_eastDirection() {
		val east = IBoard.calculateDirection(Pair(2, 2), Pair(3, 2))
		assert(east == Direction.EAST)
	}

	@Test
	fun test_southEastDirection() {
		val southEast = IBoard.calculateDirection(Pair(2, 2), Pair(3, 3))
		assert(southEast == Direction.SOUTH_EAST)
	}

	@Test
	fun test_southDirection() {
		val south = IBoard.calculateDirection(Pair(2, 2), Pair(2, 3))
		assert(south == Direction.SOUTH)
	}

	@Test
	fun test_southWestDirection() {
		val southWest = IBoard.calculateDirection(Pair(2, 2), Pair(1, 3))
		assert(southWest == Direction.SOUTH_WEST)
	}

	@Test
	fun test_westDirection() {
		val west = IBoard.calculateDirection(Pair(2, 2), Pair(1, 2))
		assert(west == Direction.WEST)
	}

	@Test
	fun test_northWestDirection() {
		val northWest = IBoard.calculateDirection(Pair(2, 2), Pair(1, 1))
		assert(northWest == Direction.NORTH_WEST)
	}

	@Test
	fun test_invalidDirection() {
		val invalid = IBoard.calculateDirection(Pair(10, 10), Pair(9, 1))
		assert(invalid == Direction.INVALID)
	}
}